python/runtime/feature_derivation/feature_derivation.py

```diff
@@ -0,0 +1,630 @@
+                                                       EmbeddingColumn,
+                                                       FeatureColumn,
+                                                       IndicatorColumn,
+                                                       NumericColumn)
+from runtime.feature_derivation.field_desc import (DataFormat, DataType,
+                                                   FieldDesc)
+from runtime.verifier import fetch_samples
+
+
+def init_column_map(target_fc_map, fc):
+    """
+    Init the target_fc_map by the feature column fc.
+
+    Args:
+        target_fc_map (dict[str -> FeatureColumn): the feature column map,
+            where the key is the field name.
+        fc (FeatureColumn): the feature column object.
+
+    Returns:
+        None.
+    """
+    if isinstance(fc, (EmbeddingColumn, IndicatorColumn)) and \
+        len(fc.get_field_desc()) == 0:
+        if fc.name not in target_fc_map:
+            target_fc_map[fc.name] = []
+
+        target_fc_map[fc.name].append(fc)
+    else:
+        for fd in fc.get_field_desc():
+            if fd.name not in target_fc_map:
+                target_fc_map[fd.name] = []
+
+            target_fc_map[fd.name].append(fc)
+
+
+def make_feature_column_map(features):
+    """
+    Build a FeatureColumn map by the features.
+
+    Args:
+        features (dict[str -> list[FeatureColumn]]): the
+            input feature columns. The key of the dict is
+            the target name, e.g. "feature_columns".
+
+    Returns:
+        A map of type dict[str -> dict[str -> list[FeatureColumn]]].
+        The key of the outer dict is the target name, e.g. "feature_columns",
+        and the key of the inner dict is the field name.
+    """
+    fc_map = {}
+    for target, fc_list in features.items():
+        if target not in fc_map:
+            fc_map[target] = {}
+
+        for fc in fc_list:
+            init_column_map(fc_map[target], fc)
+
+    return fc_map
+
+
+def make_field_desc_map(features):
+    """
+    Build a FieldDesc dict by the features.
+
+    Args:
+        features (dict[str -> list[FeatureColumn]]): the
+            input feature columns. The key of the dict is
+            the target name, e.g. "feature_columns".
+
+    Returns:
+        A map of type dict[str -> FieldDesc], where the
+        key is the field name.
+    """
+    fd_map = {}
+    for _, fc_list in features.items():
+        for fc in fc_list:
+            for fd in fc.get_field_desc():
+                fd_map[fd.name] = fd
+
+    return fd_map
+
+
+def new_default_field_desc(name):
+    """
+    Create a new default FieldDesc object.
+
+    Args:
+        name: the FieldDesc name.
+
+    Returns:
+        A FieldDesc object whose name is the given name,
+        and the data type is INT.
+    """
+    return FieldDesc(name=name, dtype=DataType.INT)
+
+
+# A regular expression to match any real number
+REAL_NUMBER_PATTERN = re.compile(
```

Comment:
Can we put this Regex expressions into a separate file?

Reply:
It is only used inside this file currently. We can remove it to a separate file when it is needed in other places in the future.